distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso

```diff
@@ -331,13 +331,39 @@ type Array
        Arguments:
        - function: A function that takes an element in the array and returns
          some transformation of that element.
+       - on_problems: Specifies how to handle any problems that arise in
+         `function`.
+
+       ? Problem Handling
+
+         The result of Errors thrown when executing the function depend on
+         `on_problems`:
+         - Report_Error: The first error is thrown.
+         - Report_Warning: The result for that element is `Nothing`, 
+           the error is attached as a warning. Currently unimplemented.
+         - Ignore: The result is `Nothing`, and the error is
+           ignored.
+
+         Errors that are thrown when executing the supplied function are wrapped
+         in `Map_Error`, which contains the index at which the error occurred.
+
+         In the `Problem_Behavior.Report_Warning` case, only `MAX_MAP_WARNINGS`
+         warnings are attached to result values. After that, the warnings are
+         dropped, but a count of the additional warnings is attached in an
+         `Additional_Warnings` warning.
 
        > Example
          Add 1 to each element of the array.
 
              [1, 2, 3].to_array . map +1
-    map : (Any -> Any) -> Vector Any
-    map self function = Vector.map self function
+    map : (Any -> Any) -> Problem_Behavior -> Vector Any
+    map self function on_problems=Problem_Behavior.Report_Error = Vector.map self function on_problems
+
+    ## PRIVATE
+       Implementation of `map` that does not do error wrapping, and does not
+       take an `on_problems` parameter.
+    map_primitive : (Any -> Any) -> Vector Any
+    map_primitive self function = Vector.map_primitive self function
```

Comment:
What is a difference between `map_primitive` and `map on_problems=Report_Error`? I believe:
```ruby
map_primitive self function = map self function Problem_Behavior.Report_Error
```
and I really don't understand why there should be two different methods to begin with!?

Reply:
No, there is a difference.

Let me show by example:
```
f = if x == 3 then Error.throw (Illegal_Argument.Error "foo") else 100+x

v = [1, 2, 3, 4, 5]
r1 = v.map_primitive f
r2 = v.map f Problem_Behavior.Report_Error
```

Now the results are following:
```
r1 = Error: Illegal_Argument.Error "foo"
r2 = Error: Map_Error.Error 2 (Illegal_Argument.Error "foo")
```

The difference is that the user-facing `map` is wrapping the original error in a wrapper that allows our users to tell at which position in the collection the error appeared. This was a common issue that when the user was e.g. downloading 100 CSVs and the 87th download has failed - all they'd get is the failure but they would not know where it happened and the result of all other downloads was discarded.

Now with `Report_Warning` - all succeeding downloads are still kept, since the error value is replaced with `Nothing` so e.g. the user can decide to skip the problematic file.

And with `Report_Error` the first error is thrown, but the user has additional information on where it has happened, hopefully making it easier to analyze what went wrong.

The separate `map_primitive` variant is needed and cannot be private, because for 'lower-level' developers (e.g. not GUI end users but library devs), it often does not make sense to wrap the error in `Map_Error` - if the `map`ping operation is an internal implementation detail, e.g. checking a list of preconditions, I want the raw error to be thrown to the user, not a wrapped one - because I do not want to 'leak' the fact that I have used the `map` internally - it is not important to the user and in that case the `Map_Error` wrapping would only add noise.

And so this `map_primitive` functionality cannot be private, because we may very well want to use it in our other libraries; as well as even our more advanced users may want to use it in their internal code.

-----

That is why we need both a `Report_Error/Report_Warning` modes for UX of our GUI users and a `map_primitive` mode which avoids these decorations and can retain the original APIs for lower level usage.

However, I don't think a special `map_primitive` function is the only solution to get this.

I think a viable alternative could be to instead allow a 4th mode on regular `map` - i.e. expose the `map_primitive` mode as `vec.map f on_problems=Nothing` or some other way (e.g. `vec.map f on_problems=Report_Raw_Error`) or whatever.

I think that approach is slightly cleaner as we still keep one `map` function.

OTOH the downside of this approach is slightly clashing requirements:
- we want the `on_problems` to have a default value, and the default should be oriented towards our GUI users as they are our main focus - so it'd have to be `Report_Error/Report_Warning`.
- that means that all usages in lib code would need to add `on_problems=Nothing` (or whatever syntax we choose) to the `map`. It seems slightly annoying and easy to forget to do so - so I imagine if we choose this, we will be forgetting from time to time and getting unnecessarily wrapped errors.


The advantage of `map_primitive` (I don't like the name) over the above solution is that it is e.g. much easier to find in our code all usages of it (search for `map_primitive`), versus the `on_problems` argument that is used on all kinds of functions so it may be harder to find.

Reply:
After discussion with @JaroslavTulach, I've moved the `_primitive` methods into extensions, and introduced a `No_Wrap` atom to implement them. Now all regular and primitive calls go through `Array_Like_Helpers.vector_from_function`.

Would we prefer perhaps `map_no_wrap` to `map_primitive`?

Reply:
Personally, I like the `map_no_wrap` name :) well maybe `map_without_wrapping`? Idk which is better, but both seem better than `map_primitive` whose name is carrying some additional meaning that does not really fit the use-case.